Attachment: Correct display of RTL labels

squelart opened the pull request:
#### 36e066b5f3e20eaa6fe432d0220e45a96c7ad129
<pre>
Attachment: Correct display of RTL labels
<a href="https://bugs.webkit.org/show_bug.cgi?id=259027">https://bugs.webkit.org/show_bug.cgi?id=259027</a>
rdar://108566229

Reviewed by Alan Baradlay.

To let the shadow DOM CSS correctly align and cut too-long labels, the elements need to have a `dir=&quot;auto&quot;`
attribute, and titles with RTL filenames should add an outer right-to-left mark so that the automatic
direction works as expected.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):

Canonical link: <a href="https://commits.webkit.org/265989@main">https://commits.webkit.org/265989@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5f6276ef3ba70ae8a27e37667b9a7844cd3d162

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12484 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14225 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12544 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12833 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14669 "126 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12648 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13385 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10553 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14651 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11296 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18392 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11752 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14654 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9872 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11186 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3073 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15516 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11802 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -522,13 +526,29 @@ String HTMLAttachmentElement::attachmentTitleForDisplay() const
+            popDirectionalIsolate,
+            zeroWidthSpace,
+            extension,
+            popDirectionalIsolate
```
I think this deserves some comment through example(s) explaining why we need all these isolation(s), zero width space etc (I know this patch just expands on the existing LTR behavior, but that initial commit from 2018 should have added such comments as well)

alanbaradlay commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -522,13 +526,29 @@ String HTMLAttachmentElement::attachmentTitleForDisplay() const
+    auto filename = StringView(title).left(indexOfLastDot);
+    auto extension = StringView(title).substring(indexOfLastDot);
+
+    if (isWideLayout() && Layout::TextUtil::directionForTextContent(filename) == TextDirection::RTL) {
```
using Layout namespace here is a bit of a layering violation (I know this is just a static helper). isn't there a WTF function on this?

squelart replied:
Ah yep, I see it's using `ubidi_getBaseDirection` from WTF/icu/unicode/ubidi.h, I'll use that as well...

squelart committed:
Attachment: Correct display of RTL labels
https://bugs.webkit.org/show_bug.cgi?id=259027
rdar://108566229

Reviewed by Alan Baradlay.

To let the shadow DOM CSS correctly align and cut too-long labels, the elements need to have a `dir="auto"`
attribute, and titles with RTL filenames should add an outer right-to-left mark so that the automatic
direction works as expected.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):

Canonical link: https://commits.webkit.org/265989@main
